runtime/lua/vim/lsp.lua

```diff
@@ -4,15 +4,15 @@ local lsp_rpc = require('vim.lsp.rpc')
 local protocol = require('vim.lsp.protocol')
 local util = require('vim.lsp.util')
 local sync = require('vim.lsp.sync')
-local api = vim.api
 
 local vim = vim
+local a = vim.api
 local nvim_err_writeln, nvim_buf_get_lines, nvim_command, nvim_buf_get_option, nvim_exec_autocmds =
-  vim.api.nvim_err_writeln,
-  vim.api.nvim_buf_get_lines,
-  vim.api.nvim_command,
-  vim.api.nvim_buf_get_option,
-  vim.api.nvim_exec_autocmds
+  a.nvim_err_writeln,
+  a.nvim_buf_get_lines,
+  a.nvim_command,
+  a.nvim_buf_get_option,
+  a.nvim_exec_autocmds
 local uv = vim.loop
 local tbl_isempty, tbl_extend = vim.tbl_isempty, vim.tbl_extend
 local validate = vim.validate
```

Comment:
Again, general question: Is this "local import everything" a pattern we want to encourage?

Reply:
(If so, we should do it consistently, e.g., for `vim.fs` etc. as well.)

Reply:
I guess yes, for performance reasons (see above); caching often-used functions from _other_ modules in upvalues is a LuaJIT performance suggestion (http://wiki.luajit.org/Numerical-Computing-Performance-Guide).

Reply:
In that case we should do this consistently ("Python style"); there's a bunch of direct `vim.tbl_contains`, for example. Other `nvim_foo` are extracted but only used once or twice.

Reply:
can probably be a judgement call, most code should not be on the hotpath.